### PR TITLE
Initial import

### PR DESCRIPTION
**Pull request by ewoutp**

### CODE REVIEW

**Comment by ewoutp** [comment]
RFR @zyndiecate

**Comment by ewoutp** [comment]
RFR @ZeissS

File: parallel.go [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+			// Make sure the token goes back
+			defer func() {
+				workerReturn <- token
+			}()
+
+			err := action()
+			if err != nil && errors != nil {
+				errors <- err
+			}
+		}(action, workerToken)
```
Comment (zyndiecate) [inline parallel.go]: `workerToken` can just be used through the scope. No need to pass it along as argument. There is only one `workerToken`, and it stays the same for each action.

File: parallel.go [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+	}
+
+	canceled := false
+	done := 0
+	for {
+		select {
+		// The "client" wants to cancel processing
+		case <-cancel:
```
Comment (zyndiecate) [inline parallel.go]: Does this work when `cancel` is `nil`?

File: parallel.go [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+			canceled = true
+			close(workerToken)
+			for range workerToken {
+				// drain it
```
Comment (zyndiecate) [inline parallel.go]: Based on this discussion, this is not necessary: https://github.com/giantswarm/companyd/pull/30#discussion-diff-38472825

File: parallel.go [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+	}
+
+	// Start actions
+	go Parallel(actions, n, errors, cancel)
+
+	// Wait for error channel to be closed or receive errors.
+	if err, ok := <-errors; ok {
+		// Fetch all other errors, just ignore them
+		for range errors {
+		}
```
Comment (zyndiecate) [inline parallel.go]: Not necessary.

File: parallel.go [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+}
+
+// ParallelAndWait calls parallel on the given arguments and waits for it to complete all actions.
+// It will return the first error.
+func ParallelAndWait(actions []func() error, n int, errors chan error, cancel <-chan struct{}) error {
+	if len(actions) == 0 {
+		return nil
+	}
+	if errors == nil {
+		return maskAny(errgo.WithCausef(nil, InvalidArgumentError, "errors is nil"))
```
Comment (zyndiecate) [inline parallel.go]: Why is `errors` required here? IMO we should create our own in case the caller don't care.

File: parallel.go (hunk 2 above)
Comment (ewoutp) [inline parallel.go]: Strangely enough yes. see http://play.golang.org/p/dW9y0jlnqO

File: parallel.go (hunk 5 above)
Comment (ewoutp) [inline parallel.go]: Good point. I'll change it

File: parallel.go [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+		}
+		return maskAny(err)
+	}
+
+	return nil
+}
+
+// ParallelAndWaitActions calls parallel on the given arguments and waits for it to complete all actions.
+// It will return the first error.
+func ParallelAndWaitActions(actions []func() error) error {
```
Comment (zyndiecate) [inline parallel.go]: I am not sure about that method. I would vote to remove this, and make `errors` optional in `ParallelAndWait`.

File: parallel.go (hunk 1 above)
Comment (ewoutp) [inline parallel.go]: Done

File: parallel.go (hunk 4 above)
Comment (ewoutp) [inline parallel.go]: Is that always true, whatever value of errors get passed into this?

Comment (zyndiecate) [inline parallel.go]: As long as the channel will be closed, it is fine.

File: parallel.go [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+	close(doneChan)
+	close(workerReturn)
+	if errors != nil {
+		close(errors)
+	}
+}
+
+// ParallelAndWait calls parallel on the given arguments and waits for it to complete all actions.
+// It will return the first error.
+func ParallelAndWait(actions []func() error, n int, errors chan error, cancel <-chan struct{}) error {
```
Comment (ZeissS) [inline parallel.go]: Why do we pass `errors` here if the client can't read it (it gets drained further down). That feels inconsistent.

File: parallel.go [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+
+	close(doneChan)
+	close(workerReturn)
+	if errors != nil {
+		close(errors)
+	}
+}
+
+// ParallelAndWait calls parallel on the given arguments and waits for it to complete all actions.
+// It will return the first error.
```
Comment (ZeissS) [inline parallel.go]: What happens with the remaining tasks? 
How long will it block?

File: parallel.go [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+}
+
+// ParallelAndWaitActions calls parallel on the given arguments and waits for it to complete all actions.
+// It will return the first error.
+func ParallelAndWaitActions(actions []func() error) error {
+	n := len(actions)
+	if n == 0 {
+		return nil
+	}
+	errors := make(chan error, n)
```
Comment (ZeissS) [inline parallel.go]: `ParallelAndWait` seems to create its own `errors` channel when not given. We don't need to create one ourself.

**Comment by xh3b4sd** [comment]
Cheers buddy. Long time no see. All the best! <3

**Pull request closed without merging**